README.md
![](https://raw.githubusercontent.com/balena-io-projects/balena-sound/master/images/balenaSound-logo.png)

# Bluetooth, Airplay and Spotify audio streaming for any audio device

**Starter project enabling you to add multi-room audio streaming via Bluetooth, Airplay or Spotify Connect to any old speakers or Hi-Fi using just a Raspberry Pi.**

**Features**
- **Bluetooth, Airplay and Spotify Connect support**: Stream audio from your favourite music services or directly from your smartphone/computer using bluetooth.
- **Multi-room synchronous playing**: Play perfectly synchronized audio on multiple devices all over your place.

### Hardware required

* Raspberry Pi 3A+/3B/3B+/4B/Zero W
* SD Card (we recommend 8GB Sandisk Extreme Pro)
* Power supply
* 3.5mm audio cable to the input on your speakers/Hi-Fi (usually 3.5mm or RCA). Alternatively you can use the HDMI port to get digital audio out.

**Notes** 
- The Raspberry Pi Zero cannot be used on it's own as it has no audio output. To use the Pi Zero you'll need to add something like the [Pimoroni pHAT DAC](https://shop.pimoroni.com/products/phat-dac) to go with it.
- For an extended list of device types supported please check this [link](DeviceSupport.md).

### Software required

* A download of this project (of course)
* Software to flash an SD card ([balenaEtcher](https://balena.io/etcher))
* A free [balenaCloud](https://balena.io/cloud) account
* The [balena CLI tools](https://github.com/balena-io/balena-cli/blob/master/INSTALL.md)

# Setup and use

Running this project is as simple as deploying it to a balenaCloud application; no additional configuration is required (unless you're using a DAC HAT).

### Setup the Raspberry Pi

* Sign up for or login to the [balenaCloud dashboard](https://dashboard.balena-cloud.com)
* Create an application, selecting the correct device type for your Raspberry Pi (we recommend setting the type as Raspberry Pi 1/Zero as your application will then be compatible with the Pi 1/Zero as well as all devices that were released afterward).
* Add a device to the application, enabling you to download the OS
* Flash the downloaded OS to your SD card with [balenaEtcher](https://balena.io/etcher)
* Power up the Pi and check it's online in the dashboard

### Deploy this application

* Install the [balena CLI tools](https://github.com/balena-io/balena-cli/blob/master/INSTALL.md)
* Login with `balena login`
* Download this project and from the project directory run `balena push <appName>` where `<appName>` is the name you gave your balenaCloud application in the first step.

# Usage

After the application has pushed and the device has downloaded the latest changes you're ready to go! Before starting, connect the audio output of your Pi to the AUX input on your Hi-Fi or speakers. You can also use the HDMI port for digital audio output.

Connect to your balenaSound device:
* If using Bluetooth: search for your device on your phone or laptop and pair.
* If using Airplay: select the balenaSound device from your audio output options.
* If using Spotify Connect: open Spotify and choose the balenaSound device as an alternate output.
* The `balenaSound xxxx` name is used by default, where `xxxx` will be the first 4 characters of the device ID in the balenaCloud dashboard.

If you are running in multi-room mode, when you start streaming the device you're connected to will configure itself as the `master` and will broadcast a message to all other devices within your balenaCloud application to get them in sync. **Note:** that it can take a few seconds for the system to autoconfigure the first time you stream.

Let the music play!

# Customize your balenaSound experience

You can configure some features of balenaSound by using environment variables. This can be set in the balena dashboard: navigate to dashboard -> your app -> Environment variables. Read more about environment variables [here](https://www.balena.io/docs/learn/manage/serv-vars/#fleet-environment-and-service-variables).

![Setting the device name](images/device-name-config.png)


### Change device name

By default, your device will be named `balenaSound xxxx`. This name will show within Airplay device lists, for Spotify Connect and when searching for devices using Bluetooth.
You can change this using `BLUETOOTH_DEVICE_NAME` environment variable that can be set in balena dashboard.

### Set output volumes

By default, balenaSound will set the output volume of your Raspberry Pi to 75% on the basis you can then control the volume upto the maximum from the connected bluetooth device. If you would like to override this, define the `SYSTEM_OUTPUT_VOLUME` environment variable.

Secondly, balenaSound will play connection/disconnection notification sounds at a volume of 75%. If this unsuitable, you can override this with the `CONNECTION_NOTIFY_VOLUME` environment variable.

**Note:** these variables should be defined as integer values without the `%` symbol.

### Multi-room

By default, balenaSound will start in multi-room mode. When running multi-room you can stream audio into a fleet of devices and have it play perfectly synchronized. It does not matter wether you have 2 or 100 devices, you only need them to be part of the same local network.

If you don't want to use multi-room or you only have one device, you can disable it by creating the `DISABLE_MULTI_ROOM` variable (with any value, for example: `1`).

**Note:** Multi-room requires a network router that supports IP multicast/broadcast (most modern routers do).

### Set bluetooth PIN code

By default, balenaSound bluetooth will connect using Secure Simple Pairing mode. If you would like to override this and use Legacy Mode with a PIN code you can do it by defining the `BLUETOOTH_PIN_CODE` environment variable. The PIN code must be numeric and up to six digits (1 - 999999).

**Note**: Legacy Mode is no longer allowed on [iOS](https://developer.apple.com/accessories/Accessory-Design-Guidelines.pdf) devices.

### GPIO bluetooth control

balenaSound allows you to control your mobile over bluetooth (Volume up/down) and enable bluetooth discovery. This feature is at default enabled, but can be disabled via setting the `DISABLE_BLUETOOTH_CONTROL` enviroment variable to `true`. You will have to connect some buttons to your GPIO pins to make use of this feature. We are using these pins:

![GPIO Pins](images/gpio-pins.png)

* GPIO17= Volume up
* GPIO27= Bluetooth discovery
* GPIO22= Volume down

### Bluetooth scripts

balenaSound has configurable scripts you can run on connect and disconnect bluetooth events. If you would like to activate this, set the  `BLUETOOTH_SCRIPTS` environment variable to `true`.
Sample scripts can be found on the `./bluetooth-audio/bluetooh-scripts/` directory, theses can be edited as needed.

### Spotify Connect over the internet

Spotify Connect only works with Spotify Premium accounts (due to the use of the [librespot](https://github.com/librespot-org/librespot) library).

### DAC Configuration

If you are using a DAC board, you will need to make a couple of changes to the device configuration in the balenaCloud dashboard.

* Disable the on-board audio by editing the existing `RESIN_HOST_CONFIG_dtparam` variable to set `”audio=off”`.
* Add an additional custom configuration variable called `BALENA_HOST_CONFIG_dtoverlay`. The value of this will depend on your DAC board. A table of values is available [here](DAC_configuration.md)

![DAC Configuration](images/dac-vars.png)

## Connect

* After the application has pushed and the device has downloaded the latest changes you're ready to go!
* Connect the audio output of your Pi to the AUX input on your Hi-Fi or speakers
* The `balenaSound bluetooth/airplay/spotify xxxx` name is used by default, where `xxxx` will be the first 4 characters of the device ID in the balenaCloud dashboard.
* If using Bluetooth: search for your device on your phone or laptop and pair.
* If using Airplay: select the balenaSound device from your audio output options.
* If using Spotify Connect: open Spotify and choose the balenaSound device as an alternate output.
* Let the music play!
If you have a Spotify Premium account you can stream locally without any configuration, but if you want to use Spotify Connect over the internet you will need to provide your Spotify credentials.

To enable Spotify login you can add your username/e-mail and password, which are set with two environment variables: `SPOTIFY_LOGIN` and `SPOTIFY_PASSWORD`.  

---

This project is in active development so if you have any feature requests or issues please submit them here on GitHub. PRs are welcome, too.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

# v3.0.0
## (2020-02-14)

* Fix var problem [Alex]
* Add disable automatic discovery car [Alex]

# v2.0.0
## (2020-02-12)

* multi-room: Add multi-room feature [Tomás Migone]

# v1.3.10
## (2020-02-10)

* Fix default volume value in README [Stevche Radevski]

# v1.3.9
## (2020-02-06)

* Add another supported DAC [Chris Crocker-White]

# v1.3.8
## (2020-02-05)

* Add InnoMaker DAC to DAC configuration docs [Jack Brown]

# v1.3.7
## (2020-02-03)

* Tweak service names [Chris Crocker-White]
* Add pisound dtoverlay [Chris Crocker-White]
* Add dtoverlay for Justboom DAC HAT [Chris Crocker-White]
* Add digital volume setting and lower default [Chris Crocker-White]

# v1.3.6
## (2020-02-03)

* Add name differentiation [Alex]

# v1.3.5
## (2020-02-03)

* Fix spotify password error if it has spaces [Alex]

# v1.3.4
## (2020-01-31)

* Add pHAT DAC config [Chris Crocker-White]
* Add DAC configuration docs [Peter Armstrong]

# v1.3.3
## (2020-01-30)

* meta: Update codeowners to HHR team [Matthew McGinn]

# v1.3.2
## (2020-01-07)

* readme: Clarify Spotify only works with Premium account [Heds Simons]

# v1.3.1
## (2020-01-07)

* Clarify Spotify connect usage [Chris Crocker-White]

# v1.3.0
## (2020-01-02)

* Add Spotify login [Alex]

# v1.2.0
## (2019-12-18)

* Add code for gpio control with udev and some other things [Alex]
* Correct typo (#66) [Chris Crocker-White]

# v1.1.3
## (2019-12-13)

* Add other supported streaming methods to README [Chris Crocker-White]

# v1.1.2
## (2019-11-20)

* blueetooth-audio: Remove supervisor api label since it's no longer needed. [Tomás Migone]
* Remove multicontainer instructions from readme since MC is now default for this project. [Tomás Migone]
* Update Dockerfiles to use exec form of CMD. [Tomás Migone]

# v1.1.1
## (2019-11-20)

* Fix typo in README [Alex]

# v1.1.0
## (2019-11-20)

* bluetooth: Reconnect automatically to last known device [Tomás Migone]

# v1.0.0
## (2019-11-15)

* PIN code, only allow numeric and 6 digits. Updated README.md. [Tomás Migone]
* Fix console debug messages on startup. [Tomás Migone]
* Add pin code instructions to README [Tomás Migone]
* Add pin code support (Legacy pairing mode) [Tomás Migone]

# v0.5.0
## (2019-11-15)

* Make raspotify start more clear, add default output volume and add cache [AlexProgrammerDE]

# v0.4.5
## (2019-11-04)

* meta: add tmigone as a codeowner [Matthew McGinn]

# v0.4.4
## (2019-10-10)

* Fix to make spotify container build for a Pi 4 app [Chris Crocker-White]

# v0.4.3
## (2019-10-10)

* Set the system volume to provided value in bluetooth-udev [Burhan Del Rey]

# v0.4.2
## (2019-10-08)

* Update CODEOWNERS (add @shaunmulligan) [Chris Crocker-White]

# v0.4.1
## (2019-10-07)

* Update codeowners and add PR template [Chris Crocker-White]

# v0.4.0
## (2019-10-07)

* Add Spotify Support [Alex]

# v0.3.0
## (2019-10-07)

* Add support for Airplay streaming [Chris Crocker-White]

# v0.2.1
## (2019-10-07)

* Add contribution guidelines [Chris Crocker-White]

# v0.2.0
## (2019-10-04)

* Add volume configuration options, fixes #28 [Chris Crocker-White]

# v0.1.0
## (2019-10-01)

* Add support for Pi 4 fixes #25 [Chris Crocker-White]

# v0.0.11
## (2019-08-01)

* Change base image to Debian Stretch instead of Buster [pwt]

# v0.0.10
## (2019-07-24)

* Restore disabling timeout functionality [Chris Crocker-White]

# v0.0.9
## (2019-07-22)

* Remove redundant bluealsa package [Chris Crocker-White]

# v0.0.8
## (2019-07-22)

* Remove redundant alsa-base package [Chris Crocker-White]

# v0.0.7
## (2019-07-11)

* fix. [Tim]

# v0.0.6
## (2019-07-05)

* Remove unnecessary action [Robert Vojta]
* Space before the screenshot [Robert Vojta]
* Cleaner log output [Robert Vojta]

# v0.0.5
## (2019-07-05)

* Allow setting custom name [Roman Mazur]

# v0.0.4
## (2019-07-05)

* Remove name from the header [Robert Vojta]

# v0.0.3
## (2019-07-04)

* Add credit for original bluetooth-agent starting point [Chris Crocker-White]
* Add logo [Chris Crocker-White]

# v0.0.2
## (2019-07-04)

* Add initial version [Chris Crocker-White]
* Rename start to start.sh [Robert Vojta]
* Reformat Python code [Robert Vojta]
* Remove unnecessary semicolon [Robert Vojta]
* Use printf & case instead of multiple ifs [Robert Vojta]
* Use printf instead of echo [Robert Vojta]
* Reformat README [Robert Vojta]
* Update README [Chris Crocker-White]
* Housekeeping, add README [Chris Crocker-White]
* Set bluetooth name including hostname [Chris Crocker-White]
* Set bluetooth name including hostname [Chris Crocker-White]
* Changed connect/disconnect sounds to cleaner ones [Chris Crocker-White]
* First working version! (req. balenaOS > 2.38) [Chris Crocker-White]
* Tuning - still losing pairings at this point [Chris Crocker-White]
* Changed to a single container - concept working [Chris Crocker-White]
* Initial commit [Chris Crocker-White]


bluetooth/start.sh
#!/usr/bin/env bash

if [[ -z "$BLUETOOTH_DEVICE_NAME" ]]; then
  BLUETOOTH_DEVICE_NAME=$(printf "balenaSound %s" $(hostname | cut -c -4))
fi

# Set the system volume here
SYSTEM_OUTPUT_VOLUME="${SYSTEM_OUTPUT_VOLUME:-75}"
echo $SYSTEM_OUTPUT_VOLUME > /usr/src/system_output_volume
printf "Setting output volume to %s%%\n" "$SYSTEM_OUTPUT_VOLUME"
amixer sset PCM,0 $SYSTEM_OUTPUT_VOLUME% > /dev/null &
amixer sset Digital,0 $SYSTEM_OUTPUT_VOLUME% > /dev/null &

# Set the volume of the connection notification sounds here
CONNECTION_NOTIFY_VOLUME="${CONNECTION_NOTIFY_VOLUME:-75}"
echo $CONNECTION_NOTIFY_VOLUME > /usr/src/connection_notify_volume
printf "Connection notify volume is %s%%\n" "$CONNECTION_NOTIFY_VOLUME"

# Set the discoverable timeout here
dbus-send --system --dest=org.bluez --print-reply /org/bluez/hci0 org.freedesktop.DBus.Properties.Set string:'org.bluez.Adapter1' string:'DiscoverableTimeout' variant:uint32:0 > /dev/null

printf "Restarting bluetooth service\n"
service bluetooth restart > /dev/null
sleep 2

# Redirect stdout to null, because it prints the old BT device name, which
# can be confusing and it also hides those commands from the logs as well.
printf "discoverable on\npairable on\nexit\n" | bluetoothctl > /dev/null

# Start bluetooth and audio agent
/usr/src/bluetooth-agent &

# If multi room is disabled remove audio redirect to fifo pipe
# Also remove if device is from Pi 1 family, since snapcast server is disabled by default
if [[ -n $DISABLE_MULTI_ROOM ]] || [[ $BALENA_DEVICE_TYPE == "raspberry-pi" ]]; then
  rm /root/.asoundrc
fi

sleep 2
rm -rf /var/run/bluealsa/
/usr/bin/bluealsa -i hci0 -p a2dp-sink --a2dp-volume &

hciconfig hci1 down > /dev/null 2>&1 # Disable onboard bluetooth if using a bluetooth dongle (onboard interface gets remapped to hci1) 

if [[ -z "$DISABLE_AUTO_DISCOVERY" ]]; then
  hciconfig hci0 up
fi

hciconfig hci0 name "$BLUETOOTH_DEVICE_NAME"

if ! [ -z "$BLUETOOTH_PIN_CODE" ] && [[ $BLUETOOTH_PIN_CODE -gt 1 ]] && [[ $BLUETOOTH_PIN_CODE -lt 1000000 ]]; then
  hciconfig hci0 sspmode 0  # Legacy pairing (PIN CODE)
  printf "Starting bluetooth agent in Legacy Pairing Mode - PIN CODE is \"%s\"\n" "$BLUETOOTH_PIN_CODE"
else
  hciconfig hci0 sspmode 1  # Secure Simple Pairing
  printf "Starting bluetooth agent in Secure Simple Pairing Mode (SSPM) - No PIN code provided or invalid\n"
fi

# Reconnect if there is a known device
sleep 2
if [ -f "/var/cache/bluetooth/reconnect_device" ]; then
  TRUSTED_MAC_ADDRESS=$(cat /var/cache/bluetooth/reconnect_device)
  printf "Attempting to reconnect to previous bluetooth device: %s\n" "$TRUSTED_MAC_ADDRESS"
  printf "connect %s\nexit\n" "$TRUSTED_MAC_ADDRESS" | bluetoothctl > /dev/null
fi

# Start gpio bluetooth control service
if [[ -z "$DISABLE_BLUETOOTH_CONTROL" ]]; then
  VOLUME_UP_GPIO="${VOLUME_UP_GPIO:-17}"
  echo $VOLUME_UP_GPIO > /usr/src/volume_up_gpio
  printf "Volume up button input is on GPIO$VOLUME_UP_GPIO %s%%\n"

  BLUETOOTH_GPIO="${BLUETOOTH_GPIO:-27}"
  echo $BLUETOOTH_GPIO > /usr/src/bluetooth_gpio
  printf "Bluetooth button input is on GPIO$BLUETOOTH_GPIO %s%%\n"

  VOLUME_DOWN_GPIO="${VOLUME_DOWN_GPIO:-22}"
  echo $VOLUME_DOWN_GPIO > /usr/src/volume_down_gpio
  printf "Volume down button input is on GPIO$VOLUME_DOWN_GPIO %s%%\n"
  
  bash /usr/src/bluetooth-control &
fi

sleep 2
printf "Device is discoverable as \"%s\"\n" "$BLUETOOTH_DEVICE_NAME"
exec /usr/bin/bluealsa-aplay --profile-a2dp 00:00:00:00:00:00
